pyartifactory/objects/build.py

```diff
@@ -0,0 +1,191 @@
+        return build_list
+
+    def delete(self, delete_build: BuildDeleteRequest) -> None:
+        """
+        :param delete_build: Model object containing required parameters
+        :return: None
+        """
+        try:
+            for _build_number in delete_build.buildNumbers:
+                self._get(
+                    f"api/{self._uri}/{delete_build.buildName}/{_build_number}",
+                )
+        except requests.exceptions.HTTPError as error:
+            self._raise_exception(error)
+        else:
+            # all build numbers exist
+            _del = self._post(f"api/{self._uri}/delete", json=delete_build.model_dump())
+            logger.debug("Builds %s deleted from %s", ",".join(delete_build.buildNumbers), delete_build.buildName)
+
+    def build_rename(self, build_name: str, new_build_name: str) -> None:
+        """
+        :param build_name: Build to be renamed
+        :param new_build_name: New Build name
+        :return: None
+        """
+        try:
+            self._get(
+                f"api/{self._uri}/{build_name}",
+            )
+        except requests.exceptions.HTTPError as error:
+            self._raise_exception(error)
+        else:
+            try:
+                self._post(f"api/{self._uri}/rename/{build_name}?to={new_build_name}")
+                logger.debug("Build %s successfully renamed to %s", build_name, new_build_name)
+            except requests.exceptions.HTTPError as error:
+                self._raise_exception(error)
+
+    def build_diff(self, build_name: str, build_number: str, older_build_number: str) -> BuildDiffResponse:
+        """
+        :param build_name: Build name to be compared
+        :param build_number: More recent build to be compared
+        :param older_build_number: Starting build to be compared
+        :return: BuildDiffResponse model object containing server response
+        """
+        build_name = build_name.strip("/")
+        build_number = build_number.strip("/")
+        older_build_number = older_build_number.strip("/")
+
+        try:
+            response = self._get(
+                f"api/{self._uri}/{build_name}/{build_number}?diff={older_build_number}",
+            )
+            logger.debug("Build Diff successfully retrieved between %s and %s", build_number, older_build_number)
+        except requests.exceptions.HTTPError as error:
+            self._raise_exception(error)
+
+        return BuildDiffResponse(**response.json())
+
+    def _raise_exception(self, error: requests.exceptions.HTTPError):
+        http_response: Union[Response, None] = error.response
+        if isinstance(http_response, Response):
+            _http_error = BuildError(**http_response.json())
+            if http_response.status_code == 404:
+                raise BuildNotFoundError(_http_error.to_error_message()) from error
+            raise ArtifactoryError(_http_error.to_error_message()) from error
+        else:
```

Comment:
I don't think the else is necessary here since you said that we return an HttpError. Also, no need to have `BuildError` since it is just a duplicate of the `Response` object. I would just do something like:

 ```
def _raise_exception(self, error: requests.exceptions.HTTPError, build_name: str):
    http_response: Union[Response, None] = error.response
    if isinstance(http_response, Response) and http_response.status_code == 404:
           raise BuildNotFoundError(f"Build {build_name} not found") from error
    raise ArtifactoryError from error
```

Reply:
HttpError is catched outside and it's passed here where:
- error status and its message coming from Response are formatted and merged in a [human readable way](https://github.com/gpongelli/python-artifactory/blob/8e0ffb3eddeb58861bad45744e887a380d77c417/pyartifactory/models/build.py#L158-L148) and there's no need to pass build_name and build_number around just to have a nice error log because those data are alread into error message; this feature is not present in actual implementation
- BuildNotFound exception is raised only in case HttpError status code is 404
- if it's a different HttpError, it's raised as ArtifactoryError with message in human readable way, this is missing in your proposal
- if HTTPError is not a Response object, just raise ArtifactoryError

BuildError model could also be reused in other part of the library to have errors in human readable way, instead of passing "build_name" and "build_numer" around to format a message for exception or debug logging.

BuildNotFound exception is needed [here](https://github.com/gpongelli/python-artifactory/blob/add-builds/pyartifactory/objects/build.py#L60) to create a build only if it's absent; a real example just tested without the BuildNotFound, having create_build as follows:
```python
    def create_build(self, create_build_request: BuildCreateRequest) -> None:
        try:
            self._put(f"api/{self._uri}", json=create_build_request.model_dump())
            logging.debug(
                "Build %s in %s successfully created",
                create_build_request.number,
                create_build_request.name,
            )
        except requests.exceptions.HTTPError as error:
            self._raise_exception(error)
```
```python
    _first_create = BuildCreateRequest(name="arti-test", number="first-bl", started="2024-08-30T12:00:19.893+0300", properties={"prop-one": "11", "prop-two": "abc"}, durationMillis=12345, url="http://jfrog-test")
    art.builds.create_build(_first_create)

    _second_create = BuildCreateRequest(name="arti-test", number="first-bl", started="2024-08-30T12:00:19.893+0300", properties={"prop-one": "999", "prop-two": "xyz"}, durationMillis=12345, url="http://jfrog-test")
    art.builds.create_build(_second_create)
```

in this case, having the same **mandatory** fields (name, number, started), the second `create_build` calls will override (or update) the existing one.

now, a decision from your side is needed to understand which way should be better:
1. keep the easy `create_build` as in this comment and "reuse" it (same call, same input model) also to be able to update a build (everything is done by PUT call) -> this choice removes BuildNotFound needs
2. use the committed implementation to manage build creation and have a slightly different one to manage update build feature -> BuildNotFound must be there to create build when really absent

let me know, thanks.

Reply:
Ok got it. Let's keep it this way.